emrun.py

```diff
@@ -1677,12 +1677,18 @@ def run():
     # to support binding to port zero we must allow the server to open to socket then retrieve the final port number
     options.port = httpd.socket.getsockname()[1]
 
+  if options.browser_url:
+    url = options.browser_url
```

Comment:
Whoops, this got merged before I was able to complete my review.

What I was going to say is that it's awfully hard to process mentally all the potential modifications to the `url` variable. These modifications are scattered over many lines and embedded in conditional statements that may or may not be mutually exclusive - quite complex and error-prone! I suggest creating a function to gather all this logic in one place so that in the body of the `run` function we could have simply:

```python
url = compute_url(options)
```

Hopefully that would help prevent future bugs like the one you're fixing here.

Reply:
@davschne-unity I completely agree, I just don't have the time at the moment. I really need to finish this work as its blocking intakes from being done which are high pri. I'll be working on a proposal in Q1 for Emunittest improvements though, and this will be among several things mentioned for improvement

Reply:
Sounds good to me, Rachel. I think there are a lot of potential improvements along these lines to make this code more comprehensible (and thus more reliable), but none of it is urgent.